core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -596,10 +596,9 @@ class TransactionStateManager(brokerId: Int,
    */
   def removeTransactionsForTxnTopicPartition(partitionId: Int, coordinatorEpoch: Int): Unit = {
     val topicPartition = new TopicPartition(Topic.TRANSACTION_STATE_TOPIC_NAME, partitionId)
-    val partitionAndLeaderEpoch = TransactionPartitionAndLeaderEpoch(partitionId, coordinatorEpoch)
 
     inWriteLock(stateLock) {
-      loadingPartitions.remove(partitionAndLeaderEpoch)
```

Comment:
to confirm, this line was a noop in practice since there would never be a case where the partitionAndLeaderEpoch from L599 exists.

with an example:

 1. become leader for partition 0 -> load partition 0, epoch 1 to loadingPartitions
 2. during load, becomes follower for partition 0 epoch 2
 3. without fix: does not remove partition 0 from loading partitions / with fix: removes partition 0 from loadingPartitions

Is my understanding correct?

What's the impact without the fix?

Reply:
Yup, basically we would never remove anything here. The impact is that we keep stale partition state loaded on the broker. This isn't harmful on its own but can affect memory usage. There was also a task to prevent reloading if a partition was already loaded on a broker (think re-electing the same leader) and having this stale state could also complicate that. https://issues.apache.org/jira/browse/KAFKA-15468